Utilities/DataLoading/EdFi.LoadTools/ApiClient/ResourceToInterchangeMapProvider.cs

```diff
@@ -2,69 +2,213 @@
+        {
+            var fileTypeByFileContexts = new Dictionary<string, List<FileContext>>();
+
+            foreach (string file in Directory.GetFiles(_dataConfiguration.Folder, "*.xml", SearchOption.AllDirectories))
+            {
+                var contextPrefix = LogContext.BuildContextPrefix(Path.GetFileName(file));
+
+                var sw = new Stopwatch();
+                sw.Start();
+
+                _log.Debug($"{contextPrefix} Start processing");
+
+                FileContext fileContext = CreateFileContext(file, contextPrefix);
+
+                if (!fileContext.IsValid)
+                {
+                    _log.Warn($"{contextPrefix} Failed Validation");
+                    continue;
+                }
+
+                if (fileTypeByFileContexts.TryGetValue(fileContext.FileType, out List<FileContext> fileContexts))
+                {
+                    fileContexts.Add(fileContext);
+                }
+                else
+                {
+                    fileTypeByFileContexts[fileContext.FileType] = new List<FileContext> {fileContext};
+                }
+
+                sw.Stop();
+
+                _log.Debug($"{contextPrefix} Finished processing in {sw.Elapsed.TotalSeconds} seconds.");
+            }
+
+            return fileTypeByFileContexts;
+
+            FileContext CreateFileContext(string fileName, string contextPrefix)
+            {
+                var fileContext = new FileContext
+                {
+                    FileName = fileName,
+                    IsValid = true
+                };
+
+                var xmlReaderSettings = new XmlReaderSettings {CloseInput = true};
+
+                if (!_xsdConfiguration.DoNotValidateXml)
+                {
+                    _log.Debug($"{contextPrefix} XSD validation is enabled.");
+                    xmlReaderSettings.Schemas = _xmlSchemaSet;
+                    xmlReaderSettings.ValidationType = ValidationType.Schema;
+
+                    xmlReaderSettings.ValidationEventHandler += (s, e) =>
                     {
-                        var resource = r.ToLower();
+                        _log.Error($"{e.Message}");
+                        fileContext.IsValid = false;
+                    };
+                }
 
-                        if (!_resourceToInterchangeDictionary.TryGetValue(resource, out List<InterchangeInfo> interchangeInfoList))
+                try
+                {
+                    using (var stream = new FileStream(fileName, FileMode.Open, FileAccess.Read))
+                    {
+                        if (stream.Length == 0)
                         {
-                            interchangeInfoList = _resourceToInterchangeDictionary[resource] = new List<InterchangeInfo>();
+                            _log.Warn($"{contextPrefix} Empty file.");
+                            throw new FileFormatException($"{contextPrefix} Empty File");
```

Comment:
Why throw for these scenarios?  You're just catching it a little further down and handling it, so this isn't a good place to use exceptions.

Reply:
well i thought about that and i removed the abort early. partially because of schema validation. so i thought it would be logical do do it the same.

Reply:
i will abort earlier as i initially did.